### PR TITLE
Add top-level error labels to write concern error

### PR DESCRIPTION
**Pull request by jyemin**
This will allow the driver to properly retry writes that the server
has labeled as a RetryableWriteError.

JAVA-4244

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java [hunk 1]
```diff
@@ -264,7 +265,9 @@ public static MongoException createSpecialException(final BsonDocument response,
         } else if (isNotPrimaryError(errorCode, errorMessage)) {
             return new MongoNotPrimaryException(response, serverAddress);
         } else if (response.containsKey("writeConcernError")) {
-            return createSpecialException(response.getDocument("writeConcernError"), serverAddress, "errmsg");
+            return createSpecialException(response.getDocument("writeConcernError").clone()
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java:268]: It's pretty hacky to clone the writeConcernError document so we can append errorLabels to it, but it's the most localized fix.  Happy to entertain other ideas though

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/unified/UnifiedRetryableWritesTest.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.junit.runners.Parameterized;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.Collection;
+
+public class UnifiedRetryableWritesTest extends UnifiedReactiveStreamsTest {
```
Comment (jyemin) [inline driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/unified/UnifiedRetryableWritesTest.java:27]: These are the first unified retryable writes tests so need new runners for them.

File: driver-core/src/test/resources/unified-test-format/retryable-writes/bulkWrite-serverErrors.json [hunk 3]
```diff
@@ -0,0 +1,205 @@
+{
```
Comment (jyemin) [inline driver-core/src/test/resources/unified-test-format/retryable-writes/bulkWrite-serverErrors.json:1]: I added these two new unified tests for this, in scope of already existing DRIVERS-1385 ticket.

**Comment by jyemin** [comment]
Sharded tests are failing... Converting to draft while investigation is in progress.

File: driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java [hunk 4]
```diff
@@ -264,7 +265,16 @@ public static MongoException createSpecialException(final BsonDocument response,
         } else if (isNotPrimaryError(errorCode, errorMessage)) {
             return new MongoNotPrimaryException(response, serverAddress);
         } else if (response.containsKey("writeConcernError")) {
-            return createSpecialException(response.getDocument("writeConcernError"), serverAddress, "errmsg");
+            BsonDocument writeConcernError = response.getDocument("writeConcernError");
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java:268]: I think this is more complex than it needs to be:

```
            MongoException error = createSpecialException(response.getDocument("writeConcernError"), serverAddress, "errmsg");
            for (BsonValue errorLabel : response.getArray("errorLabels", new BsonArray())) {
                error.addLabel(errorLabel.asString().getValue());
            }
            return error;
```

**Review by rozza — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java (hunk 4 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java]: That's so much better.  Thanks.  Added a couple of checks to a conditional protecting the loop.  One is necessary for correctness, the other just to document intent.

Done.

**Review by rozza — APPROVED** [review]
LGTM

**Pull request merged**